src/Features/LanguageServer/Protocol/CustomProtocol/FindUsagesLSPContext.cs

```diff
@@ -0,0 +1,291 @@
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Classification;
+using Microsoft.CodeAnalysis.Editor;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.FindSymbols.Finders;
+using Microsoft.CodeAnalysis.FindUsages;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Microsoft.VisualStudio.Text.Adornments;
+using Roslyn.Utilities;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.CustomProtocol
+{
+    internal class FindUsagesLSPContext : FindUsagesContext
+    {
+        private const int MaxResultsChunkSize = 32;
+
+        private readonly Dictionary<DefinitionItem, int> _definitionToId =
+            new Dictionary<DefinitionItem, int>();
+
+        private readonly List<VSReferenceItem> _resultsChunk =
+            new List<VSReferenceItem>();
+
+        private readonly IProgress<object[]> _progress;
+        private readonly Document _document;
+        private readonly int _position;
+        private readonly IMetadataAsSourceFileService _metadataAsSourceFileService;
+
+        // Methods in FindUsagesLSPContext can be called by multiple threads concurrently.
+        // We need this sempahore to ensure that we aren't making concurrent
+        // modifications to data such as _id, _definitionToId, and _resultsChunk.
+        private readonly SemaphoreSlim _semaphore = new SemaphoreSlim(1);
+
+        // Unique identifier given to each definition and reference.
+        private int _id = 0;
+
+        public override CancellationToken CancellationToken { get; }
+
+        public FindUsagesLSPContext(
+            IProgress<object[]> progress,
+            Document document,
+            int position,
+            IMetadataAsSourceFileService metadataAsSourceFileService,
+            CancellationToken cancellationToken)
+        {
+            _progress = progress;
+            _document = document;
+            _position = position;
+            _metadataAsSourceFileService = metadataAsSourceFileService;
+
+            CancellationToken = cancellationToken;
+        }
+
+        public async override Task OnCompletedAsync()
+        {
+            var referencesToReport = ArrayBuilder<VSReferenceItem>.GetInstance();
+            using (await _semaphore.DisposableWaitAsync(CancellationToken).ConfigureAwait(false))
+            {
+                if (_resultsChunk.IsEmpty())
+                {
+                    referencesToReport.Free();
+                    return;
+                }
+
+                referencesToReport.AddRange(_resultsChunk.ToArray());
+                _resultsChunk.Clear();
+            }
+
+            // We can report outside of the lock here since _progress is thread-safe.
+            _progress.Report(referencesToReport.ToArrayAndFree());
```

Comment:
```suggestion
            _progress.Report(referencesToReport.ToArray());
```

or, just no ToArray, if .Report can report an IEnumerable.

Reply:
It can't report an IEnumerable, unfortunately. :(